File: components/sloked-nodejs/src/modules/cryptoSocket.ts
import { Transform, TransformOptions, TransformCallback, Duplex, DuplexOptions } from 'stream'
import { Crypto, StreamEncryption, EncryptedDuplexStream } from '../types/crypto'
import { EventEmitter } from '../modules/eventEmitter'
import Crc32 from '../modules/crc32'

enum FrameType {
    Data = 0,
    KeyChange = 1
}

class Frame {
    constructor (type: FrameType, payload: Buffer) {
        this._type = type
        this._checksum = Crc32.Calculate(payload)
        this._payload = payload
    }

    getType(): FrameType {
        return this._type
    }

    getChecksum(): number {
        return this._checksum
    }

    getPayload(): Buffer {
        return this._payload
    }

    async encrypt(crypto: Crypto, key?: Buffer): Promise<Buffer> {
        if (this._payload.length > 0) {
            const iv = crypto.RandomBytes(crypto.IVSize())
            const encrypted = key
                ? await crypto.encrypt(this._payload, key, iv)
                : this._payload
            const result: Buffer = Buffer.alloc(9 + iv.length + encrypted.length)
            const signature = await Frame.sign(this._checksum, async (buf: Buffer) => key ? crypto.encrypt(buf, key, iv) : buf)
            result.writeUInt8(this._type)
            result.writeUInt32LE(encrypted.length, 1)
            result.writeUInt32LE(signature, 5)
            iv.copy(result, 9)
            encrypted.copy(result, 9 + iv.length)
            return result;
        } else {
            return Buffer.from([this._type, 0, 0, 0, 0])
        }
    }

    static async decrypt(crypto: Crypto, input: Buffer, key?: Buffer): Promise<[Frame, number] | null> {
        const MinimalHeaderLength: number = 5;
        if (input.length < MinimalHeaderLength) {
            return null
        }
        if (typeof FrameType[input[0]] === 'undefined') {
            throw new Error("CryptoSocket: Invalid frame type");
        }
        const type: FrameType = input[0]
        const totalLength: number = input.readUInt32LE(1)
        if (totalLength > 0) {
            const EncryptedHeaderSize: number = 9 + crypto.IVSize();
            const signature: number = input.readUInt32LE(5)
            if (input.length < totalLength + EncryptedHeaderSize) {
                return null
            }

            const iv: Buffer = input.slice(EncryptedHeaderSize - crypto.IVSize(), EncryptedHeaderSize)
            const encrypted: Buffer = input.slice(EncryptedHeaderSize, EncryptedHeaderSize + totalLength)
            const raw = key
                ? await crypto.decrypt(encrypted, key, iv)
                : encrypted
            const actualSignature = await Frame.sign(Crc32.Calculate(raw), async (buf: Buffer) => key ? crypto.encrypt(buf, key, iv) : buf)
            if (actualSignature != signature) {
                throw new Error(
                    "CryptoSocket: Actual CRC32 doesn't equal to expected");
            }
            return [new Frame(type, raw), EncryptedHeaderSize + totalLength]
        } else {
            return [new Frame(type, Buffer.alloc(0)), MinimalHeaderLength]
        }
    }

    static async sign(checksum: number, encrypt: (buf: Buffer) => Promise<Buffer>): Promise<number> {
        const rawSignatureBuffer = Buffer.alloc(4)
        rawSignatureBuffer.writeUInt32LE(checksum)
        const signatureBuffer = await encrypt(rawSignatureBuffer)
        return Crc32.Calculate(signatureBuffer)   
    }

    private _type: FrameType
    private _checksum: number
    private _payload: Buffer
}

export class EncryptionStream extends Transform {
    constructor(crypto: Crypto, key?: Buffer, options?: TransformOptions) {
        super(options)
        this._crypto = crypto
        this._key = key
        this._pending = Buffer.alloc(0)
    }

    async setKey (key?: Buffer, id?: string): Promise<void> {
        if (id) {
            const frame: Frame = new Frame(FrameType.KeyChange, id !== null ? Buffer.from(id) : Buffer.alloc(0))
            const encrypted = await frame.encrypt(this._crypto, this._key)
            this._pending = Buffer.concat([this._pending, encrypted])
        }
        this._key = key
    }

    async _transform(chunk: Buffer, _: string, done: TransformCallback): Promise<void> {
        let result = Buffer.alloc(0)
        if (chunk != null) {
            result = await this._transformBlock(chunk, this._key)
        }
        result = Buffer.concat([this._pending, result])
        this._pending = Buffer.alloc(0)
        done(null, result)
    }

    private async _transformBlock(chunk: Buffer, key?: Buffer): Promise<Buffer> {
        const frame: Frame = new Frame(FrameType.Data, chunk)
        return frame.encrypt(this._crypto, key)
    }

    private _crypto: Crypto
    private _key?: Buffer
    private _pending: Buffer
}

export class DecryptionStream extends Transform {
    constructor(crypto: Crypto, keyChangeEmitter: EventEmitter<string | null>, key?: Buffer, options?: TransformOptions) {
        super(options)
        this._crypto = crypto
        this._key = key
        this._encBuffer = Buffer.alloc(0)
        this._keyChangeEmitter = keyChangeEmitter
    }

    setKey (key?: Buffer): void {
        this._key = key
    }

    _transform(chunk: Buffer, _: string, done: TransformCallback): void {
        if (chunk == null) {
            done(null, chunk)
        } else {
            this._transformData(chunk, this._key).then(result => {
                if (result) {
                    done(null, result)
                } else {
                    done(null, Buffer.alloc(0))
                }
            }, err => done(err, null))
        }
    }

    async _transformData(chunk: Buffer, key?: Buffer): Promise<Buffer | null> {
        const result = await Frame.decrypt(this._crypto, chunk, key)
        if (result !== null) {
            const [frame, length] = result
            this._encBuffer = this._encBuffer.slice(length)
            switch (frame.getType()) {
                case FrameType.Data:
                    return result[0].getPayload()
                
                case FrameType.KeyChange:
                    await this._keyChangeEmitter.emit(frame.getPayload().length > 0 ? frame.getPayload().toString() : null)
                    break
            }
        }
        return null
    }

    private _crypto: Crypto
    private _key?: Buffer
    private _encBuffer: Buffer
    private _keyChangeEmitter: EventEmitter<string | null>
}

export class CryptoStream extends EncryptedDuplexStream implements StreamEncryption {
    constructor (rawStream: Duplex, crypto: Crypto, key?: Buffer, options?: DuplexOptions) {
        super(options)
        this._raw = rawStream
        this._buffer = []
        this._autopush = false
        this._keyChangeEmitter = new EventEmitter<string | null>()
        this._key = key
        this._in = new DecryptionStream(crypto, this._keyChangeEmitter, key)
        this._out = new EncryptionStream(crypto, key)
        this._raw.pipe(this._in)
        this._out.pipe(this._raw)

        this._in.on('readable', () => {
            let chunk;
            while (null !== (chunk = this._in.read())) {
                this._buffer.push(chunk)
            }
            this.pushBuffer()
        })
        this._raw.on('pause', () => {
            this.emit('pause')
        })
        this._raw.on('resume', () => {
            this.emit('resume')
        })
        this._raw.on('error', err => {
            this.emit('error', err)
        })
        this._raw.on('end', () => {
            this.emit('end')
        })
        this._raw.on('close', () => {
            this.emit('close')
        })
    }

    getEncryption(): StreamEncryption {
        return this
    }

    setEncryptionKey(key?: Buffer, id?: string): void {
        this._in.setKey(key)
        this._out.setKey(key, id)
    }

    getEncryptionKey(): Buffer | undefined {
        return this._key
    }

    onKeyChange(listener: ((id?: string) => Promise<void>) | ((id?: string) => void)): () => void {
        return this._keyChangeEmitter.asyncSubscribe(async val => {
            if (val) {
                await listener(val)
            } else {
                await listener()
            }
        })
    }

    _write(chunk: Buffer, enc: string, callback: any): void {
        this._out.write(chunk, enc, callback)
    }

    async _read(_?: number): Promise<void> {
        this._autopush = true
        this.pushBuffer()
    }

    private pushBuffer(): void {
        let pushed: number = 0
        while (pushed < this._buffer.length && this._autopush) {
            this._autopush = this.push(this._buffer[pushed++])
        }
        this._buffer.splice(0, pushed)
    }

    private _raw: Duplex
    private _in: DecryptionStream
    private _out: EncryptionStream
    private _buffer: Buffer[]
    private _autopush: boolean
    private _key?: Buffer
    private _keyChangeEmitter: EventEmitter<string | null>
}

File: components/framework/src/net/CryptoSocket.cpp
/*
  SPDX-License-Identifier: LGPL-3.0

  Copyright (c) 2019-2020 Jevgenijs Protopopovs

  This file is part of Sloked project.

  Sloked is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License version 3 as
  published by the Free Software Foundation.


  Sloked is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with Sloked.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "sloked/net/CryptoSocket.h"

#include <cassert>
#include <iostream>

#include "sloked/core/Encoding.h"
#include "sloked/core/Error.h"
#include "sloked/core/Hash.h"

namespace sloked {

    template <typename T>
    constexpr uint8_t ByteAt(T value, std::size_t idx) {
        assert(idx < sizeof(T));
        return (value >> (idx << 3)) & 0xff;
    }

    class SlokedCryptoSocket::Frame {
     public:
        enum class Type : uint8_t { Data = 0, KeyChange };

        Frame(Type type, SlokedSpan<const uint8_t> payload)
            : type(type), checksum(SlokedCrc32::Calculate(
                              payload.Data(), payload.Data() + payload.Size())),
              payload(payload.Data(), payload.Data() + payload.Size()) {}

        Type GetType() const {
            return this->type;
        }

        SlokedCrc32::Checksum GetChecksum() const {
            return this->checksum;
        }

        const std::vector<uint8_t> &GetPayload() const {
            return this->payload;
        }

        std::vector<uint8_t> Encrypt(SlokedCrypto::Cipher &cipher,
                                     SlokedCrypto::Key &key,
                                     SlokedCrypto::Random &random) const {
            if (!this->payload.empty()) {
                std::size_t totalLength = this->payload.size();
                if (this->payload.size() % cipher.Parameters().BlockSize != 0) {
                    totalLength =
                        (this->payload.size() / cipher.Parameters().BlockSize +
                         1) *
                        cipher.Parameters().BlockSize;
                }
                auto raw = this->payload;
                raw.insert(raw.end(), totalLength - this->payload.size(), 0);
                SlokedCrypto::Data init_vector;
                for (std::size_t i = 0; i < cipher.Parameters().IVSize; i++) {
                    init_vector.push_back(random.NextByte());
                }
                std::vector<uint8_t> encrypted;
                if (!raw.empty()) {
                    encrypted = cipher.Encrypt(raw, key, init_vector);
                }
                std::vector<uint8_t> header{static_cast<uint8_t>(this->type),
                                            ByteAt(this->payload.size(), 0),
                                            ByteAt(this->payload.size(), 1),
                                            ByteAt(this->payload.size(), 2),
                                            ByteAt(this->payload.size(), 3),
                                            ByteAt(this->checksum, 0),
                                            ByteAt(this->checksum, 1),
                                            ByteAt(this->checksum, 2),
                                            ByteAt(this->checksum, 3)};
                header.insert(header.end(), init_vector.begin(),
                              init_vector.end());
                encrypted.insert(encrypted.begin(), header.begin(),
                                 header.end());
                return encrypted;
            } else {
                std::vector<uint8_t> message{static_cast<uint8_t>(this->type),
                                             0u, 0u, 0u, 0u};
                return message;
            }
        }

        static std::optional<Frame> Decrypt(std::vector<uint8_t> &input,
                                            SlokedCrypto::Cipher &cipher,
                                            SlokedCrypto::Key &key) {
            constexpr std::size_t MinimalHeaderLength = 5;
            if (input.size() < MinimalHeaderLength) {
                return std::optional<Frame>{};
            }
            if (input.at(0) > static_cast<uint8_t>(Type::KeyChange)) {
                throw SlokedError("CryptoSocket: Invalid frame type");
            }
            Type type = static_cast<Type>(input.at(0));
            const std::size_t length = input.at(1) + (input.at(2) << 8) +
                                       (input.at(3) << 16) +
                                       (input.at(4) << 24);
            if (length > 0) {
                const std::size_t EncryptedHeaderSize =
                    9 + cipher.Parameters().IVSize;
                const SlokedCrc32::Checksum crc32 =
                    input.at(5) + (input.at(6) << 8) + (input.at(7) << 16) +
                    (input.at(8) << 24);
                std::size_t totalLength = length;
                if (length % cipher.Parameters().BlockSize != 0) {
                    totalLength = (length / cipher.Parameters().BlockSize + 1) *
                                  cipher.Parameters().BlockSize;
                }
                if (input.size() < totalLength + EncryptedHeaderSize) {
                    return std::optional<Frame>{};
                }

                SlokedCrypto::Data init_vector;
                init_vector.insert(init_vector.end(),
                                   input.begin() + EncryptedHeaderSize -
                                       cipher.Parameters().IVSize,
                                   input.begin() + EncryptedHeaderSize);
                std::vector<uint8_t> encrypted(
                    input.begin() + EncryptedHeaderSize,
                    input.begin() + EncryptedHeaderSize + totalLength);
                input.erase(input.begin(),
                            input.begin() + EncryptedHeaderSize + totalLength);
                const auto raw = cipher.Decrypt(encrypted, key, init_vector);
                auto actualCrc32 =
                    SlokedCrc32::Calculate(raw.begin(), raw.begin() + length);
                if (actualCrc32 != crc32) {
                    throw SlokedError(
                        "CryptoSocket: Actual CRC32 doesn't equal to expected");
                }
                return Frame(type, SlokedSpan(raw.data(), length));
            } else {
                input.erase(input.begin(), input.begin() + MinimalHeaderLength);
                return Frame(type, SlokedSpan<const uint8_t>(nullptr, 0));
            }
        }

     private:
        Frame(Type type, SlokedCrc32::Checksum checksum,
              std::vector<uint8_t> payload)
            : type(type), checksum(checksum), payload(payload) {}
        Type type;
        SlokedCrc32::Checksum checksum;
        std::vector<uint8_t> payload;
    };

    SlokedCryptoSocket::SlokedCryptoSocket(
        std::unique_ptr<SlokedSocket> socket,
        std::unique_ptr<SlokedCrypto::Cipher> cipher,
        std::unique_ptr<SlokedCrypto::Key> key,
        std::unique_ptr<SlokedCrypto::Random> random)
        : socket(std::move(socket)),
          cipher(std::move(cipher)), defaultKey{nullptr}, key(std::move(key)),
          random(std::move(random)) {}

    SlokedCryptoSocket::SlokedCryptoSocket(SlokedCryptoSocket &&socket)
        : socket(std::move(socket.socket)), cipher(std::move(socket.cipher)),
          defaultKey(std::move(socket.defaultKey)), key(std::move(socket.key)),
          encryptedBuffer(std::move(socket.encryptedBuffer)),
          buffer(std::move(socket.buffer)) {}

    SlokedCryptoSocket &SlokedCryptoSocket::operator=(
        SlokedCryptoSocket &&socket) {
        this->socket = std::move(socket.socket);
        this->cipher = std::move(socket.cipher);
        this->defaultKey = std::move(socket.defaultKey);
        this->key = std::move(socket.key);
        this->encryptedBuffer = std::move(socket.encryptedBuffer);
        this->buffer = std::move(socket.buffer);
        return *this;
    }

    bool SlokedCryptoSocket::Valid() {
        return this->socket != nullptr && this->socket->Valid();
    }

    void SlokedCryptoSocket::Close() {
        if (this->socket) {
            this->socket->Close();
            this->socket = nullptr;
            this->cipher = nullptr;
        }
    }

    std::size_t SlokedCryptoSocket::Available() {
        if (this->Valid()) {
            this->Fetch();
            return this->buffer.size();
        } else {
            throw SlokedError("CryptoSocket: Invalid socket");
        }
    }

    bool SlokedCryptoSocket::Closed() {
        if (this->Valid()) {
            this->Fetch();
            return this->socket->Closed();
        } else {
            throw SlokedError("CryptoSocket: Invalid socket");
        }
    }

    bool SlokedCryptoSocket::Wait(std::chrono::system_clock::duration timeout) {
        if (this->Valid()) {
            if (this->buffer.empty()) {
                this->socket->Wait(timeout);
            }
            this->Fetch();
            return this->buffer.size();
        } else {
            return false;
        }
    }

    std::optional<uint8_t> SlokedCryptoSocket::Read() {
        if (this->Valid()) {
            this->Fetch();
            if (!this->buffer.empty()) {
                auto res = this->buffer.at(0);
                this->buffer.erase(this->buffer.begin());
                return res;
            } else {
                return {};
            }
        } else {
            throw SlokedError("CryptoSocket: Invalid socket");
        }
    }

    std::vector<uint8_t> SlokedCryptoSocket::Read(std::size_t count) {
        if (this->Valid()) {
            this->Fetch(count);
            std::vector<uint8_t> msg(this->buffer.begin(),
                                     this->buffer.begin() + count);
            this->buffer.erase(this->buffer.begin(),
                               this->buffer.begin() + count);
            return msg;
        } else {
            throw SlokedError("CryptoSocket: Invalid socket");
        }
    }

    void SlokedCryptoSocket::Write(SlokedSpan<const uint8_t> data) {
        if (this->Valid()) {
            this->Put(data.Data(), data.Size());
        } else {
            throw SlokedError("CryptoSocket: Invalid socket");
        }
    }

    void SlokedCryptoSocket::Write(uint8_t data) {
        if (this->Valid()) {
            this->Put(&data, 1);
        } else {
            throw SlokedError("CryptoSocket: Invalid socket");
        }
    }

    void SlokedCryptoSocket::Flush() {
        if (this->Valid()) {
            this->socket->Flush();
        } else {
            throw SlokedError("CryptoSocket: Invalid socket");
        }
    }

    std::unique_ptr<SlokedIOAwaitable> SlokedCryptoSocket::Awaitable() const {
        return this->socket->Awaitable();
    }

    SlokedSocketEncryption *SlokedCryptoSocket::GetEncryption() {
        return this;
    }

    void SlokedCryptoSocket::SetEncryption(
        std::unique_ptr<SlokedCrypto::Key> key,
        std::optional<std::string> keyId) {

        this->SendKeyChangeNofitication(keyId);
        if (this->defaultKey == nullptr) {
            this->defaultKey = std::move(this->key);
        }
        this->key = std::move(key);
    }

    void SlokedCryptoSocket::RestoreDefaultEncryption(
        std::optional<std::string> keyId) {
        this->SendKeyChangeNofitication(keyId);
        if (this->defaultKey) {
            this->key = std::move(this->defaultKey);
        }
    }

    std::function<void()> SlokedCryptoSocket::OnKeyChange(
        std::function<void(const std::optional<std::string> &)> listener) {
        return this->keyChangeEmitter.Listen(std::move(listener));
    }

    void SlokedCryptoSocket::SendKeyChangeNofitication(
        const std::optional<std::string> &keyId) {
        if (keyId.has_value()) {
            EncodingConverter conv(Encoding::Get("system"), Encoding::Utf8);
            auto encodedKeyId = conv.Convert(keyId.value());
            Frame frame(
                Frame::Type::KeyChange,
                SlokedSpan<const uint8_t>(
                    reinterpret_cast<const uint8_t *>(encodedKeyId.c_str()),
                    keyId->size()));
            auto encrypted =
                frame.Encrypt(*this->cipher, *this->key, *this->random);
            this->socket->Write(SlokedSpan(encrypted.data(), encrypted.size()));
            this->Flush();
        }
    }

    void SlokedCryptoSocket::Put(const uint8_t *bytes, std::size_t length) {
        Frame frame(Frame::Type::Data, SlokedSpan(bytes, length));
        auto encrypted =
            frame.Encrypt(*this->cipher, *this->key, *this->random);
        this->socket->Write(SlokedSpan(encrypted.data(), encrypted.size()));
    }

    void SlokedCryptoSocket::Fetch(std::size_t sz) {
        do {
            auto chunk = this->socket->Read(this->socket->Available());
            this->encryptedBuffer.insert(this->encryptedBuffer.end(),
                                         chunk.begin(), chunk.end());
            auto frame = Frame::Decrypt(this->encryptedBuffer, *this->cipher,
                                        *this->key);
            if (frame.has_value()) {
                this->InsertFrame(frame.value());
            }
        } while (this->buffer.size() < sz || this->socket->Available() > 0);
    }

    void SlokedCryptoSocket::InsertFrame(const Frame &frame) {
        switch (frame.GetType()) {
            case Frame::Type::Data:
                if (!frame.GetPayload().empty()) {
                    this->buffer.insert(this->buffer.end(),
                                        frame.GetPayload().begin(),
                                        frame.GetPayload().end());
                }
                break;

            case Frame::Type::KeyChange: {
                std::optional<std::string> param;
                if (!frame.GetPayload().empty()) {
                    EncodingConverter conv(Encoding::Utf8,
                                           Encoding::Get("system"));
                    param = conv.Convert(std::string(frame.GetPayload().begin(),
                                                     frame.GetPayload().end()));
                }
                this->keyChangeEmitter.Emit(param);
            } break;
        }
    }

    SlokedCryptoServerSocket::SlokedCryptoServerSocket(
        std::unique_ptr<SlokedServerSocket> serverSocket, SlokedCrypto &crypto,
        std::unique_ptr<SlokedCrypto::Key> key)
        : serverSocket(std::move(serverSocket)), crypto(crypto),
          defaultKey(std::move(key)) {}

    SlokedCryptoServerSocket::SlokedCryptoServerSocket(
        SlokedCryptoServerSocket &&serverSocket)
        : serverSocket(std::move(serverSocket.serverSocket)),
          crypto(serverSocket.crypto),
          defaultKey(std::move(serverSocket.defaultKey)) {}

    bool SlokedCryptoServerSocket::Valid() {
        return this->serverSocket != nullptr && this->serverSocket->Valid();
    }

    void SlokedCryptoServerSocket::Start() {
        if (this->Valid()) {
            this->serverSocket->Start();
        } else {
            throw SlokedError("CryptoServerSocket: Invalid socket");
        }
    }

    void SlokedCryptoServerSocket::Close() {
        if (this->serverSocket) {
            this->serverSocket->Close();
            this->serverSocket = nullptr;
        }
    }

    std::unique_ptr<SlokedSocket> SlokedCryptoServerSocket::Accept(
        std::chrono::system_clock::duration timeout) {
        if (this->Valid()) {
            auto rawSocket = this->serverSocket->Accept(timeout);
            if (rawSocket) {
                return std::make_unique<SlokedCryptoSocket>(
                    std::move(rawSocket), this->crypto.NewCipher(),
                    this->defaultKey->Clone(), this->crypto.NewRandom());
            } else {
                return nullptr;
            }
        } else {
            throw SlokedError("CryptoServerSocket: Invalid socket");
        }
    }

    std::unique_ptr<SlokedIOAwaitable> SlokedCryptoServerSocket::Awaitable()
        const {
        return this->serverSocket->Awaitable();
    }

    SlokedCryptoSocketFactory::SlokedCryptoSocketFactory(
        SlokedSocketFactory &socketFactory, SlokedCrypto &crypto,
        std::unique_ptr<SlokedCrypto::Key> key)
        : socketFactory(socketFactory), crypto(crypto),
          defaultKey(std::move(key)) {}

    SlokedCryptoSocketFactory::SlokedCryptoSocketFactory(
        SlokedCryptoSocketFactory &&socketFactory)
        : socketFactory(socketFactory.socketFactory),
          crypto(socketFactory.crypto),
          defaultKey(std::move(socketFactory.defaultKey)) {}

    std::unique_ptr<SlokedSocket> SlokedCryptoSocketFactory::Connect(
        const SlokedSocketAddress &addr) {
        auto rawSocket = this->socketFactory.Connect(addr);
        if (rawSocket) {
            return std::make_unique<SlokedCryptoSocket>(
                std::move(rawSocket), this->crypto.NewCipher(),
                this->defaultKey->Clone(), this->crypto.NewRandom());
        } else {
            return nullptr;
        }
    }

    std::unique_ptr<SlokedServerSocket> SlokedCryptoSocketFactory::Bind(
        const SlokedSocketAddress &addr) {
        auto rawSocket = this->socketFactory.Bind(addr);
        if (rawSocket) {
            return std::make_unique<SlokedCryptoServerSocket>(
                std::move(rawSocket), this->crypto, this->defaultKey->Clone());
        } else {
            return nullptr;
        }
    }
}  // namespace sloked